### PR TITLE
delete lo network when TearDownPod to avoid CNI cache leak

### PR DESCRIPTION
**Pull request by mars1024**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Libcni in vendor has been updated to v0.7.1, it will [cache some results](https://github.com/kubernetes/kubernetes/blob/f1453953b703a0cb8cb47dbdfde4c9be846f8350/vendor/github.com/containernetworking/cni/libcni/api.go#L263) on disk during adding CNI network, if we don't delete CNI network, the cache files will be left on disk forever which will cause leak problem, such as running out of iNode.

Maybe in the future, libcni will do some improvements on this, but we need to delete lo network when TearDownPod to avoid CNI cache leak now.

**Special notes for your reviewer**:
[libcni release v0.7.0](https://github.com/containernetworking/cni/releases/tag/v0.7.0)

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/dockershim/network/cni/cni.go [hunk 1]
```diff
@@ -328,6 +328,13 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 		klog.Warningf("CNI failed to retrieve network namespace path: %v", err)
 	}
 
+	// Windows doesn't have loNetwork. It comes only with Linux
+	if plugin.loNetwork != nil {
+		if err = plugin.deleteFromNetwork(plugin.loNetwork, name, namespace, id, netnsPath, nil); err != nil {
+			return err
```
Comment (mattjmcnaughton) [inline pkg/kubelet/dockershim/network/cni/cni.go]: Currently, if we fail to delete the `loNetwork`, we won't attempt to delete the `getDefaultNetwork`. Is that expected?

Or should we aggregate errors and return them all at the end?

**Review by mattjmcnaughton — COMMENTED** [review]
/ok-to-test

Thanks for your pr!

How difficult would it be to add unit tests for this behavior?

Would also love someone from sig-network to take a look.

**Comment by dcbw** [comment]
@mars1024 Just ran into the same thing today in ocicni while adding use-cache-on-DEL :) Good catch.

File: pkg/kubelet/dockershim/network/cni/cni.go (hunk 1 above)
Comment (dcbw) [reply pkg/kubelet/dockershim/network/cni/cni.go]: @mars1024 I would say just ignore the failure. the loopback network doesn't actually need any cleanup, so we can ignore errors here and they shouldn't be fatal.

Comment (mars1024) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Thanks @dcbw  @mattjmcnaughton , will update this!

**Comment by mars1024** [comment]
PR updated, please go on reviewing! @dcbw @mattjmcnaughton

**Review by mattjmcnaughton — CHANGES_REQUESTED** [review]
Thanks for the update :)

I think the test failures are legitimate. Mind taking a look/fixing them?

File: pkg/kubelet/dockershim/network/cni/cni.go [hunk 2]
```diff
@@ -328,6 +328,12 @@ func (plugin *cniNetworkPlugin) TearDownPod(namespace string, name string, id ku
 		klog.Warningf("CNI failed to retrieve network namespace path: %v", err)
 	}
 
+	// Windows doesn't have loNetwork. It comes only with Linux
+	if plugin.loNetwork != nil {
+		// Loopback network deletion failure should not be fatal on teardown
+		_ = plugin.deleteFromNetwork(plugin.loNetwork, name, namespace, id, netnsPath, nil)
```
[the remark this answers is not in the document]
Comment (mars1024) [reply pkg/kubelet/dockershim/network/cni/cni.go]: Thanks @mattjmcnaughton , but it seems a failure about mock **DelNetworkList** and another about nil pointer, I'm trying to fix them.

**Comment by mars1024** [comment]
CI failures resolved! Please go on reviewing, maybe we should speed up to get this merged before k/k 1.16 code freeze. Thanks a lot ! @mattjmcnaughton @dcbw

**Pull request merged**